File: java/abc177/C.java
import java.util.*;

public class C {

  private static final long div = 1_000_000_007;

  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int n = sc.nextInt();
    long[] a = new long[n];
    long[] acc = new long[n];
    Arrays.setAll(a, i -> sc.nextInt());
    for (int i = n - 1; i >= 0; i--) {
      if (i == n - 1) {
        acc[i] = 0;
      } else {
        acc[i] = (acc[i + 1] + a[i + 1]) % div;
      }
    }

    // System.out.println(Arrays.toString(acc));

    long sum = 0;
    for (int i = 0; i < n; i++) {
      sum += (a[i] * acc[i]) % div;
      sum = sum % div;
    }
    System.out.println(sum);

  }

}

File: rust/abc177/src/c.rs
use proconio::input;
use std::collections::VecDeque;
use std::io::{stdout, BufWriter, Write};

const div: usize = 1_000_000_000 + 7;

pub fn main() {
  let out = stdout();
  let mut out = BufWriter::new(out.lock());
  input! {n: usize};
  input! {a: [usize; n]}

  let mut acc: VecDeque<usize> = VecDeque::new();
  for idx in (0..n).rev() {
    if idx == n - 1 {
      acc.push_front(0);
    } else {
      let val = &acc.get(0).unwrap();
      acc.push_front((**val + a.get(idx + 1).unwrap()) % div);
    }
  }

  let sum: usize = acc
    .iter()
    .enumerate()
    .fold(0, |acc, x| (acc + (x.1 * a.get(x.0).unwrap()) % div) % div);
  writeln!(out, "{}", sum % div).unwrap();
}
